Add short/long date formats, time formats, temperature and speed unit settings

Montellese opened the pull request:
These commits basically add temperature and speed unit settings to `Appearance` -> `International` which can be set independently of the `Region` setting. The idea is that the default setting is `regional` which will pick whatever is specified in the configured region but it is also possible to choose another unit from a list of supported units.

Here's a screenshot of the new settings:
![Screenshot](https://www.dropbox.com/s/90tt1ctgywpomup/Screenshot%202015-03-08%2002.15.37.png?dl=1)

Apart from the commits adding/improving `CTemperature`, `CSpeed`, `CLangInfo` and the settings logic I also added a commit (see 588e5c233b28191cc5d210f7b029a8013e45857f) that refactors `CLangCodeExpander` by basically renaming most of the methods to something more obvious using the proper ISO 639-X names.

Montellese committed:
move Temperature.h/cpp to utils

Montellese committed:
CTemperature: cleanup and refactoring

Montellese committed:
settings: add locale.temperatureunit setting

Montellese committed:
CLangCodeExpander: cleanup and refactoring

Montellese committed:
utils: add CSpeed utility class

Montellese committed:
CLangInfo: refactor speed unit handling using CSpeed

Montellese committed:
settings: add locale.speedunit setting

mkortstiege commented:
Looks good. I actually wonder if we need this in core since usage is pretty much limited to weather addons but i guess moving this to yet another resource addon would be not worth the hassle.

Montellese commented:
Well my plan is to move the region definitions into resource addons at one point (but not for 15.0) but IMO these units are pretty much fixed and shoulnd't really change over time. The only thing I was wondering is if the unit settings should be in the weather setting category since they are only used there. But then the connection to the selected region is not so obvious anymore.

mkortstiege commented:
Ok, sounds good. I'd keep it in there as it is common to have region, currency and units in one spot.

MartijnKaijser commented:
as mentioned on IRC yesterday, I noticed we never used the official naming of speed units.  kmh -> km/h
and others

mkortstiege commented:
@Montellese please pick https://github.com/mkortstiege/xbmc/commit/e593ba9611f46bd25a0fd5097acb84c4de87564e

da-anda commented:
nice. One thing that came to my mind that could also be added at some point would be the preferred time (am/pm + 12h, 24h) and date formats (dd.mm.yy, dd.mm.yyyy, mm/dd/yy, ...).
Not related to this PR but something I noticed recently is that we don't automatically set the timezone according to the selected region (or something like that) - could be an OE only issue though.

Montellese commented:
@da-anda: I'm currently working on time format and 12/24-hour clock format settings. Date format will come after that.

Concerning timezones I'm not sure either how that actually works. On non-win32 platforms you should be able to choose the timezone separately.

MartijnKaijser commented:
is timezone/time format something we could fetch from the OS and still show it?
Not having this one windows causes some issues when using JSON iirc (read some thread that had problems because it was missing)

Montellese commented:
@mkortstiege: Thanks for the xcode sync. I integrated it into the respective commits. But didn't you miss removing `Temperature.h/cpp` from the main directory?
jenkins build this please

mkortstiege commented:
Most likely yes. If it breaks just nuke the old refs or i will update once at home.

Paxxi commented:
@Montellese before you spend too much time on this, most or all of this data we can grab from icu as long as we have a region. If you want to explore it it's separate from the charset stuff I'm working on so can be added on its own.

ace20022 commented:
@Paxxi I would love to help integrating icu into the langcodeexpander in the near future. I'll ping you on irc.

Montellese commented:
@Paxxi: We still need to integrate it into our settings and almost all of the stuff is already there. I just need to change the way we access and store the values chosen by the user. Changing where the values are defined can still be easily done later but the integration as settings can be done independent of that.

Montellese commented:
jenkins build this please

Montellese commented:
Hm something didn't work. jenkins build this please

Montellese committed:
settings: add locale.timeformat and locale.use24hourclock settings

Montellese committed:
settings: add locale.shortdateformat and locale.longdateformat settings

Montellese commented:
OK I added two more commits that add settings for
* short date format
* long date format
* time format
* whether to use 12-hour or 24-hour clock format

I extracted all the different formats for each of these settings from the existing information in the `langinfo.xml` files of the different languages.

Montellese commented:
I updated the screenshot in the PR description to show all newly added settings.

Montellese commented:
Let's see if jenkins is willing to work again. jenkins build this please

Montellese commented:
jenkins build this please

ace20022 commented:
@Montellese Apart from my comment I'm okay, nice refactor!

Montellese commented:
@ace20022: Thanks for the review. I have addressed your comment and rebased to master. Any objections to this being merged? jenkins build this please

Memphiz commented:
test if i can get the unstable build result passed through to the PR jenkins build this please

Memphiz commented:
one last try - we can only choose between unstable (yellow builds) beeing treated as "build failed" or "build success" - i guess you guys want it to be treated as "build success" - jenkins build this please

da-anda commented:
I just noticed that when you change the setting "Use 12/24-hour format" that this also affects the "Time format" setting above. I suppose that this is required for am/pm changes. Anyhow the "Time format" won't change back to "regional" if it was like this before you changed the 12/24-h setting.
I also wondered why we have some regions twice, once with 12h and once with 24h when 12/24h can be configured right below.

Montellese commented:
> I just noticed that when you change the setting "Use 12/24-hour format" that this also affects the "Time format" setting above. I suppose that this is required for am/pm changes.

Yes that's correct. Changing the 12/24-hour format automatically changes the time format because you have to choose one with/without AM/PM.

> Anyhow the "Time format" won't change back to "regional" if you switch the 12/24 back to "regional" and the "time format" was set to regional before.

Yes, I've noticed that too but there's no way to know what the previous setting was. No matter which way you do it (fall back to regional or to a specific value) you will always get cases where the procedure is "wrong".

> I also wondered why we have some regions twice, once with 12h and once with 24h when 12/24h can be configured right below.

I kept them to be backwards compatible to the old settings. But obviously we could remove/merge those and add some update logic that would try to figure out what the merged region is based on the old value.
